src/server/qgsserver.cpp

```diff
@@ -482,14 +482,25 @@ void QgsServer::handleRequest( QgsServerRequest &request, QgsServerResponse &res
 
         // Dispatcher: if SERVICE is set, we assume a OWS service, if not, let's try an API
         // TODO: QGIS 4 fix the OWS services and treat them as APIs
-        QgsServerApi *api = nullptr;
-
         if ( !response.feedback() || !response.feedback()->isCanceled() )
         {
-          if ( params.service().isEmpty() && ( api = sServiceRegistry->apiForRequest( request ) ) )
+          if ( params.service().isEmpty() )
           {
-            const QgsServerApiContext context { api->rootPath(), &request, &responseDecorator, project, sServerInterface };
-            api->executeRequest( context );
+            QgsServerApi *api = sServiceRegistry->apiForRequest( request );
+            if ( api )
+            {
+              const QgsServerApiContext context { api->rootPath(), &request, &responseDecorator, project, sServerInterface };
+              api->executeRequest( context );
+            }
+            else
+            {
+              throw QgsOgcServiceException(
```

Comment:
Looks god to me, just wondering if is there a way we can print the actual list of available services and APIs instead of an hardcoded message that might not necessarily reflect the status of the server.

Reply:
Imho, I would rather consider a minimal  explanatory error message  and get the list of services/apis in an appropriate landing-page.

It appears also that there is no public method in `QgsServiceRegistry` for iterating over registered apis/services - and may be that would require an appropiate PR.